File: Leetcode Daily Challenge/027-September-2025/015-maximum-number-of-words-you-can-type.cpp
// https://leetcode.com/problems/maximum-number-of-words-you-can-type/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // TC : O(M + N)
    // SC : O(1)
    int canBeTypedWords(string text, string brokenLetters) {
        vector<int> brokenFreq(26, 0);
        for (char &ch : brokenLetters) {
            if (!brokenFreq[ch - 'a']) 
                brokenFreq[ch - 'a'] = 1;
        }

        istringstream ss(text);
        string word;
        int result = 0;

        while (getline(ss, word, ' ')) {
            bool canType = true;
            for (char &ch : word) {
                if (brokenFreq[ch - 'a']) {
                    canType = false;
                    break;
                }
            }
            if (canType) result += 1;
        }

        return result;
    }
};

File: Leetcode Daily Challenge/027-September-2025/015-maximum-number-of-words-you-can-type.java
// Leetcode Daily Challenge/027-September-2025/015-maximum-number-of-words-you-can-type.cpp/




import java.util.*;

class Solution {
    // TC : O(M + N)
    // SC : O(1)
    public int canBeTypedWords(String text, String brokenLetters) {
        int[] brokenFreq = new int[26];
        for (char ch : brokenLetters.toCharArray()) {
            if (brokenFreq[ch - 'a'] == 0) {
                brokenFreq[ch - 'a'] = 1;
            }
        }

        boolean isBroken = false;
        int result = 0;

        for (char ch : text.toCharArray()) {
            if (ch == ' ') {
                if (isBroken == false) {
                    result += 1;
                }
                isBroken = false;
            } else if (brokenFreq[ch - 'a'] == 1) {
                isBroken = true;
            }
        }

        if (isBroken == false) result += 1;

        return result;
    }
}